README.md

```diff
@@ -71,10 +71,146 @@ origen
 The driver contains a number of registers which are written to directly by the VPI process, allowing it to drive or expect a given data value (stored in <code>origen.pins.\<pin\>.data</code>) by writing a 1 to <code>origen.pins.\<pin\>.drive</code> or <code>origen.pins.\<pin\>.compare respectively</code>.
 If the value being driven by the pin does match the expect data during a cycle, then an error signal will be asserted by the driver and this will increment an error counter that lives in <code>origen.debug.errors[31:0]</code>.
 
+### Toolchains
 
+Running the testbench along with the VPI takes place within the toolchain. Different toolchains have different setups
+and different compilation and running procedures. Below is summary of some of the supported toolchains and notes on how
+to use them with OrigenSim.
+
+<anchor>cadence</anchor>
+<anchor>irun</anchor>
+#### Cadence (irun)
+
+#### Synopsis
+
+#### Generic
+
+Generic toolchains allow you to use a tool that is not support out of the box by <code>OrigenSim</code>. For these, it 
+is your responsiblity, using the <code>pre_run_start_block</code> and the
+<code>post_run_start_block</code> to start the VPI process, however, this allows for arbitrary commands to be run, 
+within the context of the simulation, and allow end users to still use <code>origen g</code>
+as if with an <code>OrigenSim</code> supported toolchain.
+
+An example of such a setup could be:
+
+~~~ruby
+OrigenSim.generic do |sim|
+  # Set a 5 minute connection timeout
+  sim.startup_timeout 300
+
+  sim.generic_run_cmd do |s|
+    # Return the command to start the testbench.
+    "path/to/custom/sim/script +socket+#{s.socket_id}"
+  end
+end
+~~~
+
+An example using the predecessor of the support Cadence tool <code>irun</code>, <code>ncsim</code> is shown below.
+
+~~~ruby
+OrigenSim.generic(startup_timeout: 900) do |sim|
+  sim.testbench_top 'na_origen'
+  sim.generic_run_cmd do |s|
+    "ncsim na_origen -loadpli origen.so:bootstrap +socket+#{s.socket_id}"
```

Comment:
na_origen refers to a custom testbench name I think, maybe should add a comment above this explaining that